compiler/noirc_evaluator/src/ssa/opt/assert_constant.rs

```diff
@@ -108,27 +113,71 @@ fn evaluate_static_assert(
     instruction: InstructionId,
     arguments: &[ValueId],
 ) -> Result<bool, RuntimeError> {
-    if arguments.len() != 2 {
+    if arguments.len() < 2 {
         panic!("ICE: static_assert called with wrong number of arguments")
     }
 
-    if !function.dfg.is_constant(arguments[1]) {
-        let call_stack = function.dfg.get_instruction_call_stack(instruction);
-        return Err(RuntimeError::StaticAssertDynamicMessage { call_stack });
+    // All of the arguments representing the message must be constants
+    for arg in arguments.iter().skip(1) {
+        if !function.dfg.is_constant(*arg) {
+            let call_stack = function.dfg.get_instruction_call_stack(instruction);
+            return Err(RuntimeError::StaticAssertDynamicMessage { call_stack });
+        }
     }
 
     if function.dfg.is_constant_true(arguments[0]) {
-        Ok(false)
-    } else {
-        let call_stack = function.dfg.get_instruction_call_stack(instruction);
-        if function.dfg.is_constant(arguments[0]) {
-            let message = function
-                .dfg
-                .get_string(arguments[1])
-                .expect("Expected second argument to be a string");
+        return Ok(false);
+    }
+
+    let call_stack = function.dfg.get_instruction_call_stack(instruction);
+    if !function.dfg.is_constant(arguments[0]) {
+        return Err(RuntimeError::StaticAssertDynamicPredicate { call_stack });
+    }
+
+    // To turn the arguments into a string we do the same as we'd do if the arguments
+    // were passed to the built-in foreign call "print" functions.
+    let mut foreign_call_params = Vec::with_capacity(arguments.len() - 1);
+    for arg in arguments.iter().skip(1) {
+        if !function.dfg.is_constant(*arg) {
+            let call_stack = function.dfg.get_instruction_call_stack(instruction);
+            return Err(RuntimeError::StaticAssertDynamicMessage { call_stack });
+        }
+        append_foreign_call_param(*arg, &function.dfg, &mut foreign_call_params);
+    }
+
+    match PrintableValueDisplay::<FieldElement>::try_from_params(&foreign_call_params) {
+        Ok(display_values) => {
+            let message = display_values.to_string();
             Err(RuntimeError::StaticAssertFailed { message, call_stack })
-        } else {
-            Err(RuntimeError::StaticAssertDynamicPredicate { call_stack })
         }
+        Err(err) => {
+            let message = match err {
```

Comment:
I think we should panic on ICE. Does this return a normal error or is it turned into a panic later?

Reply:
Ah, I saw `RuntimeError::InternalError` already has some variants which produce "ICE: ..." as a message, that's why I copied that pattern. But it should be fine to use panics here. Should I?

Reply:
Yes, please use panics for ICE so that it's easier to detect them when testing, e.g. ICE will only be found by fuzzers when it panics.

Reply:
Done!